src/clients/memcache.rs

```diff
@@ -0,0 +1,297 @@
+                    continue;
+                }
+            }
+        }
+
+        let mut s = stream.take().unwrap();
+
+        let work_item = work_receiver
+            .recv()
+            .await
+            .map_err(|_| Error::new(ErrorKind::Other, "channel closed"))?;
+
+        REQUEST.increment();
+
+        // check if we should reconnect
+        if work_item == WorkItem::Reconnect {
+            CONNECT_CURR.sub(1);
+            continue;
+        }
+
+        let request = Request::try_from(&work_item);
+
+        // skip unsupported work items
+        if request.is_err() {
+            stream = Some(s);
+            continue;
+        }
+
+        // compose request
+        REQUEST_OK.increment();
+        request.unwrap().compose(&mut write_buffer);
+
+        // send request
+        let start = Instant::now();
+        s.write_all(write_buffer.borrow()).await?;
+
+        // clear the buffers
+        write_buffer.clear();
+        read_buffer.clear();
+
+        // read until response or timeout
+        let response = loop {
+            let remaining_time = config
+                .client()
+                .unwrap()
+                .request_timeout()
+                .as_millis()
+                .saturating_sub(start.elapsed().as_millis().into());
+            if remaining_time == 0 {
+                break Err(ResponseError::Timeout);
+            }
+
+            match timeout(
+                Duration::from_millis(remaining_time as _),
+                s.read(read_buffer.borrow_mut()),
+            )
+            .await
+            {
+                Ok(Ok(n)) => {
+                    unsafe {
+                        read_buffer.advance_mut(n);
+                    }
+                    match parser.parse(read_buffer.borrow()) {
+                        Ok(resp) => {
+                            let consumed = resp.consumed();
+                            let resp = resp.into_inner();
+
+                            read_buffer.advance(consumed);
+
+                            break Ok(resp);
+                        }
+                        Err(e) => match e.kind() {
+                            ErrorKind::WouldBlock => {}
+                            _ => {
+                                break Err(ResponseError::Exception);
+                            }
+                        },
+                    }
+                }
+                Ok(Err(_)) => {
+                    break Err(ResponseError::Exception);
+                }
+                Err(_) => {
+                    break Err(ResponseError::Timeout);
+                }
+            }
+        };
+
+        let stop = Instant::now();
+
+        match response {
+            Ok(response) => {
+                let latency_ns = stop.duration_since(start).as_nanos();
+
+                if validate_response(&work_item, &response).is_err() {
+                    RESPONSE_EX.increment();
+                    CONNECT_CURR.sub(1);
```

Comment:
What has happened to the connection here?

Reply:
Added comments and reorganized a bit so that it should be clearer. Please take a look.